libcst/codegen/transforms.py

```diff
@@ -0,0 +1,18 @@
+import libcst as cst
+
+
+class SimplifyUnionsTransformer(cst.CSTTransformer):
+    def leave_Subscript(
+        self, original_node: cst.Subscript, updated_node: cst.Subscript
+    ) -> cst.BaseExpression:
+        if updated_node.value.deep_equals(cst.Name("Union")):
+            slc = updated_node.slice
+            if isinstance(slc, cst.Slice):
+                raise Exception("Unexpected Slice in Union!")
+            elif isinstance(slc, cst.Index):
+                return slc.value
+            elif len(slc) == 1:
+                # This is a Union[SimpleType] which is equivalent to
+                # just SimpleType
```

Comment:
You're right about this, and I should update the comment. In the spot it was copied from, it could be that it was previously modified in the tree and only one `ExtSlice` was left, since I wasn't rendering out between passes. This is one particular spot in the grammar where either an `Index` or a single `ExtSlice` can render to the same thing. "Fixing" that would make it hard to generate code by adding `ExtSlice` one by one though, and we will never parse to a single `ExtSlice` so I think its fine to just update the comment here.

Reply:
On a tangent, it makes me wonder why indexing with a tuple is modeled as ExtSlice thing rather than just an index that is a tuple? At runtime it's just indexing with a tuple:

```
>>> class Foo:
...     def __getitem__(self, idx):
...             print(idx)
...
>>> f = Foo()
>>> f[1, 2, 3]
(1, 2, 3)
```

But I would guess there are good reasons for this that I'm not thinking of.

Reply:
I legitimately do not know either. I bet the answer is buried in one of the PEPs somewhere. Its weird that there would be this specific exemption for `ExtSlice` when it was added for numerical analysis (and used for typing) and nothing else in Python supports it.

Reply:
Hmm, this seems to be a LibCST specific thing. In Python's AST, `f[1, 2, 3]` parses to a Subscript node whose `.slice` attribute is an `Index` whose `.value` is `Tuple`:

```
>>> t = ast.parse("f[1, 2, 3]")
>>> t.body[0].value
<_ast.Subscript object at 0x7efeb70b08d0>
>>> t.body[0].value.slice
<_ast.Index object at 0x7efeb70b0978>
>>> t.body[0].value.slice.value
<_ast.Tuple object at 0x7efeb70b09b0>
```

Whereas in LibCST, it parses to a list of `ExtSlice`.

In Python's AST, I think the extended form of slice is only for when you have slice ranges, like `1:3`; it's not related to the use of commas/tuples in a slice.

Maybe there is a good reason for this difference, but it strikes me as possibly a bug/mistake, and maybe one that would clean up some code like this if we fixed it (and if it were gonna be fixed, sooner would be better than later.)

@bgw do you have any additional context here? If it is a bug, I can file a separate issue (and maybe work on a PR too).

Reply:
Hmm, it strikes me that this would end up making it even harder to work with `Subscript`, as you would need to be able to recognize `Slice`, `Index`, `ExtSlice` *and* `Tuple`. I think what you're proposing is technically correct, but it would mean that dealing with `Subscript` becomes even more irregular than it already is which is something we desperately tried to avoid in LibCST.

Reply:
The grammar specifies the following (simplified to remove other trailer types):

```
trailer: '[' subscriptlist ']'
subscriptlist: subscript (',' subscript)* [',']
subscript: test | [test] ':' [test] [sliceop]
sliceop: ':' [test]
```

Which implies that the fact the AST parses a `subscriptlist` into a tuple is an implementation detail for CPython. It doesn't actually specify `ExtSlice` anywhere. I think the mistake here is in allowing a `Slice`, `Index` or `Sequence[Extslice]` instead of having it ALWAYS be an `ExtSlice` which represents `subscriptlist`.

Reply:
This would also simplify `Subscript` in a similar manner, because you're always guaranteed a sequence, and would get rid of the need to check for sequence versus node types.

Reply:
Interesting. I guess that special-casing of comma-separated lists in the grammar is necessary to allow range-slices within tuple elements, which otherwise would not be allowed. That must be the thing that was added to support some numerical use cases, since I think using a regular tuple as an index has been allowed for a long time (the presence of `test` in the grammar means any expression can be an index).

Given that it's definitely more reasonable to have this as a special form, I guess.

FWIW though my proposal (and matching AST) would have nested the `Tuple` inside the `Index`, where you can already have any expression at all. So it wouldn't increase the irregularity. It would decrease it, because the `.slice` of a Subscript would never be a list of anything, it would always either be an `Index` or a `Slice`.

Reply:
Ok, the AST does parse `f[1, 2:3]` into an `ExtSlice`, because grammatically tuples can't contain a Slice. So the question here is just whether the case of `f[1, 2, 3]` should be made consistent with `f[(1, 2, 3)]` or consistent with `f[1, 2, 3:4]`. AST chooses the former, LibCST currently chooses the latter. I don't think either one has an advantage in overall regularity, and I think it makes slightly more sense for the unusual node type to be triggered by the presence of the otherwise-illegal range-slice, not by the presence of commas. But not sure it's worth changing at this point.

Reply:
It comes down to a question of possibility of deprecation. It should be easy to modify LibCST to always parse to a `Sequence[ExtSlice]` going forward since the types allow it and existing code will not be surprised by its existing. Then we can mark a deprecation date where we no longer support construction with a raw `Index` or `Slice` and close the typing gap. Going towards tuple instead would be more difficult.

Reply:
I summarized our conversation here in https://github.com/Instagram/LibCST/issues/101